src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -266,6 +266,10 @@ bool CircularBoundedJournal::is_passed_end(scan_valid_records_cursor &cursor,
       return print_invalid(r_header);
     }
   } else if (r_header.committed_to.segment_seq != cursor.seq.segment_seq) {
+    if (r_header.committed_to == cursor.last_committed &&
+	r_header.committed_to.segment_seq == cursor.seq.segment_seq - 1) {
```

Comment:
IIUC this is assuming that if cursor.last_committed is updated, it must point to the current segment_seq.

But I don't think the above assumption is correct, for example the below case should be reasonable:
* When start to write the first rolled record-group, the last third record-group before-rolling is committed;
* When start to write the second rolled record-group, the last second is committed, so committed-to is still pointing to the last sequence after the update.

Reply:
Hm.. probably, I guess I don't understand your example. Am I missing something else?

Assuming that there are three record_group_header with data in a batch manner at the near end of the journal 
(like a unit test I added).
In this example, all record_group_header's committed_to is identical, right?
At replay() time, the cursor's last_committed will be the same because they (the record_group_headers) have
the same committed_to, and Replay() increases journal_seq_t (seq), including segment_seq, when it tries to read at the beginning of the journal (rolled)---this is done by where it reads. Therefore, we can identify whether these records are written at once or not---if the record_group_header has the same committed_to, we can be sure that they are processed simultaneously.
Can you check a unit test I added?

Reply:
I think in case of the unit test, writing to the memory is not asynchronous, so submit_record() with writing will be completed immediately, or at least we cannot effectively enumerate how writes are delayed.

For example there are 6 record-groups (a~f) to submit with rolling, in the following order:
1. submit_record(a);
2. submit_record(b);
3. submit_record(c);
4. roll to begin
5. submit_record(d);
6. submit_record(e);
7. submit_record(f);

When starts to submit(d) at step 5, submit(a) has completed its finalize phase [1], so the header of d's committed_to points to a.
When starts to submit(e) at step 6, submit(b) has completed its finalize phase [1], so the header of e's committed_to points to b.

I think that would break the above logic during replay, marking e as invalid incorrectly.

Is that reasonable?

[1] https://github.com/ceph/ceph/blob/f55ad812263234cae9c621991e0edc7569b14b69/src/crimson/os/seastore/journal/circular_bounded_journal.cc#L104-L107

Reply:
Understood. How about adding ">=" replacing with "==" to detect that case, as below?
```
bool is_record_segment_seq_valid() {
...
   } else if (r_header.committed_to.segment_seq != cursor.seq.segment_seq) {
+    if (r_header.committed_to >= cursor.last_committed &&
+       r_header.committed_to.segment_seq == cursor.seq.segment_seq - 1) {
+      return true;
+    }
     return print_invalid(r_header);
   } 
...
```

Reply:
See the comparator implemented in journal_seq_t, I assume we should only compare the offset part of the committed_to.

Reply:
Sorry for wrong suggestion. r_header.committed_to.offset is right.